DOC-3356: New feature documentation for `tinymceai`. (#3976)
tinymce/tinymce-docs: modules/ROOT/pages/tinymceai-models.adoc

== How Model Selection Works

The agent model (`agent-1`) automatically selects the best underlying model based on:

* **Request complexity**: Simple queries use faster, cost-effective models
* **Content size**: Larger content uses models with better context windows
* **Required capabilities**: Web search and reasoning require compatible models
* **Cost optimization**: Balances quality with cost efficiency

[[model-configuration]]
=== Model Configuration

Models can be configured in two ways:

* **Plugin Configuration**: Use `tinymceai_default_model` and `tinymceai_allow_model_selection` options. See xref:tinymceai.adoc#options-chat[Chat configuration options] for full details.
* **API Configuration**: Specify the model directly in API requests for more control. See xref:tinymceai-models.adoc#api-examples[API Examples] below.

.Example: Set default model in plugin configuration
[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to the HTML
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_default_model: 'gemini-2-5-flash',
  tinymceai_allow_model_selection: true,

>   tinymceai_token_provider: () => fetch('/api/token').then(r => r.json())

Reviewer: I am not sure about using this form for the publicly-visible examples. It's concise, but 
```suggestion
  tinymceai_token_provider: () => fetch('/api/token').then(r => r.text()).then(token => ({ token }));
```
... would show the response structure more clearly. I know most integrations would return a JSON object from their endpoint, but we need to convey that `tinymceai_token_provider` expects an object of a specific format to be returned.

Author: Resolved in [this commit](https://github.com/tinymce/tinymce-docs/pull/3976/changes/5c0e531cc97658bbe2c3e53a8959983362d057d1) thanks @tjdett

= TinyMCE AI Models

:navtitle: AI Models
:pluginname: TinyMCE AI
:plugincode: tinymceai
:description: AI model configuration for TinyMCE AI
:description_short: AI model configuration
:keywords: AI, models, configuration, tinymceai

> {pluginname} supports multiple AI models from different providers. Model selection and configuration can be done through the plugin UI or through the REST API. This page covers model capabilities, configuration options, and limitations that apply to both plugin and API usage.

Reviewer 2: ```suggestion
{pluginname} supports multiple AI models from different providers. Model selection and configuration can be done through the plugin configuration and UI or through the REST API. This page covers model capabilities, configuration options, and limitations that apply to both plugin and API usage.
```

[[recommended-agent-models]]
== Recommended: Agent Models

The `agent-1` model automatically selects the best AI model for requests based on speed, quality, and cost. It is the recommended choice for most use cases as it optimizes performance and cost automatically.

[[available-models]]

> == Available Models

Reviewer 3: Missing models that are available to select in AI chat: 'gpt-5.4', 'claude-4-6-sonnet', 'gemini-3-1-pro'

Author: Thanks @soritaheng updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3976/commits/6559a0439641e92cfde4642260bc107dfc8cfb2a)

[cols="1,2,1,1,2"]
|===
|Model |Description |xref:tinymceai-chat.adoc#web-search[Web Search] |xref:tinymceai-chat.adoc#reasoning[Reasoning] |Configuration id

|**Auto (default)**
|Automatically selects best model for speed, quality, and cost.
|Yes
|Yes
|`'auto'` (also `'agent-1'`, learn more about xref:tinymceai-models.adoc#model-compatibility-versions[compatibility versions])

|**GPT-5.2**
|OpenAI's flagship model for advanced reasoning, creativity, and complex tasks
|Yes
|Yes
|`'gpt-5.2'`

|**GPT-5.1**
|OpenAI's flagship model for advanced reasoning, creativity, and complex tasks
|Yes
|Yes
|`'gpt-5.1'`

|**GPT-5**
|OpenAI's flagship model for advanced reasoning, creativity, and complex tasks
|Yes
|Yes
|`'gpt-5'`

|**GPT-5 Mini**
|A lightweight version of GPT-5: faster, more cost-efficient
|Yes
|Yes
|`'gpt-5-mini'`

|**Claude 4.5 Haiku**
|Cost-efficient model for quick interactions with improved reasoning
|Yes
|Yes
|`'claude-4-5-haiku'`

|**Claude 4.5 Sonnet**
|Advanced model with improved creativity, reliability, and reasoning
|Yes
|Yes
|`'claude-4-5-sonnet'`

|**Gemini 3 Pro**
|Google's advanced model for versatile problem-solving and research
|Yes
|Yes

> |`'gemini-3-pro'`

Reviewer 3: ```suggestion
```
There is no 'gemini-3-pro' in the available models' list.

Author: Updated table based on fresh call to  `GET /v1/models/{version}`  ref: https://github.com/tinymce/tinymce-docs/pull/3976/commits/775a3a816f565570eeb761e8d498b928f068bf20

|**Gemini 3 Flash**
|Lightweight Gemini model for fast, cost-efficient interactions
|Yes
|Yes
|`'gemini-3-flash'`

|**Gemini 2.5 Flash**
|Lightweight Gemini model for fast, cost-efficient interactions
|Yes
|Yes
|`'gemini-2-5-flash'`

|**GPT-4.1**
|OpenAI's model for reliable reasoning, speed, and versatility
|Yes
|No
|`'gpt-4.1'`

|**GPT-4.1 Mini**
|A lighter variant of GPT-4.1 that balances speed and cost while maintaining solid accuracy
|Yes
|No
|`'gpt-4.1-mini'`
|===

[[limitations]]
=== Limitations

Model availability depends on the subscription tier and service compatibility version. Some models may have specific limitations or requirements.

[[how-model-selection-works]]
== How Model Selection Works

The agent model (`agent-1`) automatically selects the best underlying model based on:

* **Request complexity**: Simple queries use faster, cost-effective models
* **Content size**: Larger content uses models with better context windows
* **Required capabilities**: Web search and reasoning require compatible models
* **Cost optimization**: Balances quality with cost efficiency

[[model-configuration]]
=== Model Configuration

> Models can be configured in two ways:

Reviewer 2: Actually the key one is in the JWT permissions, because that affects which models are allowed to be used in both the plugin and API. The plugin config and API variables just let you pick which of the allowed models are used in certain situations. Same with the variables when setting up custom actions and such which let you pick a model for that...

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3976/commits/842336e560bf51ceb0da227601bb293a74514792)

Also included a opening paragraph that states that JWT model permissions decides which models may be used, and that the plugin UI, API bodies, and per-action settings only choose among allowed models (with Quick Actions as an example).

> * **Plugin Configuration**: Use `tinymceai_default_model` and `tinymceai_allow_model_selection` options. See xref:tinymceai.adoc#options-chat[Chat configuration options] for full details.

Reviewer 2: ```suggestion
* **Plugin Configuration**: Use `tinymceai_default_model` and `tinymceai_allow_model_selection` options to configure which models are available to users in the plugin UI. See xref:tinymceai.adoc#options-chat[Chat configuration options] for full details.
```

* **API Configuration**: Specify the model directly in API requests for more control. See xref:tinymceai-models.adoc#api-examples[API Examples] below.

[NOTE]
====

> The set of models and capabilities can change between major {productname} releases. That kind of change is not treated as a breaking change for the editor. Use the plugin model picker or `GET /v1/models/{version}` to see what the environment offers and the model limitations.

Reviewer 2: "plugin model picker"?

Author: LMAO: maybe the wrong word here.
Suggestion: Update to be `model selector`?
```suggestion
The set of models and capabilities can change between major {productname} releases. That kind of change is not treated as a breaking change for the editor. Use the model selector or `GET /v1/models/{version}` to see what the environment offers and the model limitations.
```

Reviewer 4: ```suggestion
The set of models and capabilities can change between major {productname} releases. That kind of change is not treated as a breaking change for the editor. Use the plugin's model selection dropdown or `GET /v1/models/{version}` to see what the environment offers and the model limitations.
```

====

.Example: Set default model in plugin configuration
[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to the HTML
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_default_model: 'gemini-2-5-flash',
  tinymceai_allow_model_selection: true,
  tinymceai_token_provider: () => fetch('/api/token').then(r => r.text()).then(token => ({ token }))
});
----

> If the token endpoint returns JSON (for example `{"token":"..."}`), resolve the body with `response.json()` and return an object that includes a `token` string property in the shape `tinymceai_token_provider` expects.

Reviewer 2: does this need to be here?

```suggestion
```

Author: Removed in [this commit](https://github.com/tinymce/tinymce-docs/pull/3976/commits/e8e318b04ae7644413ae2afbb69965386f4d5d10)

**Always-on reasoning models:**

* Reasoning is always active during inference.
* Reasoning cannot be turned off through the API.

To determine if a model has always-on reasoning, check the API response when listing models or refer to the model capabilities in the plugin UI. Models with mandatory reasoning will indicate this in their capability structure.

NOTE: Model names such as `gpt-5`, `claude-4-sonnet`, and similar are examples. Actual available models depend on the service compatibility version. Use the `/v1/models` API endpoint or check the plugin model selection dropdown to see current available models for the environment.

[[web-scraping]]
=== Web Scraping

Web scraping extracts and processes content from web pages so the AI can analyze and summarize it. When users add web resources as context in xref:tinymceai-chat.adoc#adding-context-to-conversations[Chat], the service fetches and parses the page content for the AI to use. Web scraping supports standard web pages and is subject to xref:tinymceai-limits.adoc#rate-limits[rate limits] for web resource requests.

[[model-limitations]]
== Model Limitations

> Per-model caps (context length, attachment sizes, PDF page totals, and similar) are returned in the `limits` object for each entry in `GET /v1/models/{version}`. Those values are **authoritative for the environment** and can differ by model (for example stricter `maxImageSize` than `maxFileSize` for some providers). See xref:tinymceai-models.adoc#verifying-model-limits-for-integration[Verifying limits for a configured model] for how to match the integration’s model id to the correct `items[]` entry.

Reviewer 2: "authoritative for the environment"? I'm not sure what this means?

Author: The limits the service applies at runtime ect. Ill update the wording.

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3976/commits/74a0b55ad3a61e15e5b35c2954c06710beca2fbf)

The sections below cover moderation, descriptions, and deprecation. Attachment limits are documented only under xref:tinymceai-models.adoc#file-processing-limits[File processing limits], using the live models API so values stay aligned with the service.

[[file-processing-limits]]
=== File Processing Limits

> {pluginname} supports common attachment types in Chat conversations, including PDF, DOCX, images, Markdown, HTML, and plain text. Per-file and per-conversation ceilings—including maximum sizes, attachment counts, and PDF page totals—are returned per model in the `limits` object from `GET /v1/models/{version}`. Those numbers are authoritative for the deployment; they can change with the service and vary by model, so read the `limits` object from that response at runtime for each model the integration uses. Field names, units, and schema updates are defined in the https://tinymceai.api.tiny.cloud/docs#tag/Models[Models API] OpenAPI documentation.

Reviewer 2: "authoritative for the deployment"? I'm not sure what this means?

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3976/commits/74a0b55ad3a61e15e5b35c2954c06710beca2fbf)
